src/core/components/pin.js

```diff
@@ -0,0 +1,589 @@
+  }
+
+  const pin = {
+    types: {
+      direct: 'direct',
+      recursive: 'recursive',
+      indirect: 'indirect',
+      internal: 'internal',
+      all: 'all'
+    },
+
+    clear: () => {
+      directPins.clear()
+      recursivePins.clear()
+      internalPins.clear()
+    },
+
+    set: pinSet(dag),
+
+    add: promisify((hashes, options, callback) => {
+      // callback (err, pinset)
+      let recursive = true  // default
+      if (typeof options === 'function') {
+        callback = options
+      }
+      else if (options && options.recursive === false) {
+        recursive = false
+      }
+      normalizeHashes(hashes, (err, mhs) => {
+        if (err) { return callback(err) }
+        const result = {
+          // async result queue
+          payload: [],
+          update: (item) => {
+            result.payload.push(item)
+            if (result.payload.length === mhs.length) {
+              pin.flush((err, root) => {
+                if (err) { return callback(err) }
+                return callback(null, result.payload)
+              })
+            }
+          }
+        }
+        mhs.forEach((multihash) => {
+          if (recursive) {
+            if (recursivePins.has(multihash)) {
+              // it's already pinned recursively
+              result.update({
+                hash: keyString(multihash),
+                type: pin.types.recursive
+              })
+              return
+            }
+
+            // recursive pin should replace direct pin
+            directPins.delete(multihash)
+
+            // entire graph of nested links should be
+            // pinned, so make sure we have all the objects
+            getRecursive(multihash, (err) => {
+              if (err) { return callback(err) }
+              // found all objects, we can add the pin
+              recursivePins.add(multihash)
+              result.update({
+                hash: keyString(multihash),
+                type: pin.types.recursive
+              })
+            })
+          } else {
+            if (recursivePins.has(multihash)) {
+              // recursive supersedes direct, can't have both
+              return callback(
+                `${keyString(multihash)} already pinned recursively`
+              )
+            }
+            if (directPins.has(multihash)) {
+              // already directly pinned
+              result.update({
+                hash: keyString(multihash),
+                type: pin.types.direct
+              })
+              return
+            }
+            // make sure we have the object
+            dag.get(new CID(multihash), (err, res) => {
+              if (err) { return callback(err) }
+              // found the object, we can add the pin
+              directPins.add(multihash)
+              result.update({
+                hash: keyString(multihash),
+                type: pin.types.direct
+              })
+            })
+          }
+        })
+      })
+    }),
```

Comment:
I think I know the crux of the problem, it is the mix between callbacks and promises and lack of calling the callbacks only once.

I believe if you change your use of promises to callbacks in the CLI code and HTTP API you will see how your code doesn't finish.